yoyodyne/dataconfig.py

```diff
@@ -0,0 +1,158 @@
+           List[str]: symbols from that cell.
+        """
+        cell = row[col - 1]  # -1 because we're using one-based indexing.
+        return list(cell) if not sep else cell.split(sep)
+
+    # Source is always present.
+
+    @property
+    def has_targets(self) -> bool:
+        return self.target_col != 0
+
+    @property
+    def has_features(self) -> bool:
+        return self.features_col != 0
+
+    # Iterators over the data; just provide a filename path.
+
+    def source_samples(self, filename: str) -> Iterator[List[str]]:
+        """Yields source."""
+        with open(filename, "r") as source:
+            tsv_reader = csv.reader(source, delimiter="\t")
+            for row in tsv_reader:
+                yield self._get_cell(row, self.source_col, self.source_sep)
+
+    def source_target_samples(
+        self, filename: str
+    ) -> Iterator[Tuple[List[str], List[str]]]:
+        """Yields source and target."""
+        with open(filename, "r") as source:
+            tsv_reader = csv.reader(source, delimiter="\t")
+            for row in tsv_reader:
+                source = self._get_cell(row, self.source_col, self.source_sep)
+                target = self._get_cell(row, self.target_col, self.target_sep)
+                yield source, target
+
+    def source_features_target_samples(
+        self, filename: str
+    ) -> Iterator[Tuple[List[str], List[str], List[str]]]:
+        """Yields source, features, and target."""
+        with open(filename, "r") as source:
+            tsv_reader = csv.reader(source, delimiter="\t")
+            for row in tsv_reader:
+                source = self._get_cell(row, self.source_col, self.source_sep)
+                features = self._get_cell(
+                    row, self.features_col, self.features_sep
+                )
+                target = self._get_cell(row, self.target_col, self.target_sep)
+                yield source, features, target
+
+    def samples(self, filename: str) -> Iterator[Tuple[List[str], ...]]:
+        """Picks the right one for this config."""
+        if self.has_features:
+            return (
+                self.source_features_target_samples(filename)
+                if self.has_targets
+                else self.source_features_samples(filename)
+            )
+        else:
+            return (
+                self.source_target_samples(filename)
+                if self.has_targets
+                else self.source_samples(filename)
+            )
+
+    # Writer help.
+
+    def row(
+        self,
+        source: List[str],
+        target: List[str],
+        features: Optional[List[str]] = None,
+    ) -> List[str]:
+        """Returns a TSV-style row using the config.
+
+        Args:
+            source (List[str]).
+            target (List[str]).
+            features (List[str], optional).
+
+        Returns:
+            List[str].
+        """
+        row = [""] * max(self.source_col, self.target_col, self.features_col)
```

Comment:
Possible user error here: since we don't limit the columns to only three, there's an event someone specifies `self.features_col = 5` and this will be jumbled. This could come up if someone wanted to use a more verbose tsv for yoyodyne.

Reply:
I consider what it does here to be the best thing you could do in that scenario---it inserts blank columns to pad so you get what you asked for. Do you think it should error out instead?

Reply:
Since you're assuming `targets` it's already going to be length 2 at minimum. So something along the lines of 
```
    row = ["",""]
    if features is not none:
         row.extend("")
```
Or something along those lines. This will remove the need to assume anything about the columns (which shouldn't influence any writing behavior anyhow). 

I also think `row` should be `_row` since it's more of a sub function only intended for writing to file.

Reply:
Discussed offline. I think this is operating as intended, @bonham79 has some alternative proposals to work out, going to table this but he may put together a PR or request-for-comments for discussion.

Reply:
That's what Travis wanted to do too, it sounds like. I'm not convinced it wouldn't be done easier in pure UNIX if someone wants that. But I'll be overruled if you both feel this way.